File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.py
from string import ascii_uppercase
def solution(msg):
    _dict = {}
    index = 1
    answer = []
    # 1. 사전 초기화
    for i in ascii_uppercase:
        _dict[i] = index
        index += 1
    
    start, end = 0, 0
    while 1:
        end += 1
        if end == len(msg):  # 종료
            answer.append(_dict[msg[start:end]])
            return answer
        
        print(msg[start:end+1])
        if not _dict.get(msg[start:end+1]):
            _dict[msg[start:end+1]] = index
            index += 1
            answer.append(_dict[msg[start:end]])
            start = end

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
const solution = msg => {
    const alphabetDict = {}
    const answer = []
    let index = 0
    for (let i="A".charCodeAt(); i<="Z".charCodeAt(); i++) {
        alphabetDict[String.fromCharCode(i)] = ++index
    }
    
    for (let i=0; i<msg.length; i++) {
        let j = i
        while (j < msg.length) {
            if (!(msg.slice(i, j + 1) in alphabetDict)) break
            j++
        }
        
        answer.push(alphabetDict[msg.slice(i, j)])
        alphabetDict[msg.slice(i, j + 1)] = ++index
        i = j - 1
    }
    return answer
}